interfaces/builtin/gpio_chardev.go

```diff
@@ -0,0 +1,213 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/systemd"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// TODO: Snapd should validate the correctness of slot declarations
+// (i.e. lines across slots are unique) when installing the gadget
+// snap e.g. when validating snap.yaml
+
+// The interface operates as follows:
+//   - uses snap-gpio-helper to set up a virtual GPIO device exposing specific
+//     lines defined in the slot as character device node at
+//     /dev/snap/gpio-chardev/<slot-snap>/<slot-name>
+//   - sets up a symlink at /dev/snap/gpio-chardev/<plug-snap>/<plug-name>
+//     pointing at the virtual slot device
+//   - slot devices are given a specific udev tag, which consumer rules match on
+//
+// https://docs.kernel.org/userspace-api/gpio/chardev.html
+// https://docs.kernel.org/admin-guide/gpio/gpio-aggregator.html
+const gpioChardevSummary = `allows access to specific GPIO chardev lines`
+
+const gpioChardevBaseDeclarationSlots = `
+  gpio-chardev:
+    allow-installation:
+      slot-snap-type:
+        - gadget
+    deny-auto-connection: true
+`
+
+var gpioChardevConnectedSlotKmod = []string{
+	"gpio-aggregator",
+}
+
+type gpioChardevInterface struct {
+	commonInterface
+}
+
+func validateSourceChips(sourceChip []string) error {
+	if len(sourceChip) == 0 {
+		return errors.New(`"source-chip" must contain at least one chip`)
+	}
+	exists := make(map[string]bool)
+	for _, chip := range sourceChip {
+		if chip == "" {
+			return errors.New(`chip in "source-chip" cannot be empty`)
+		}
+		if chip != strings.TrimSpace(chip) {
+			return fmt.Errorf(`chip in "source-chip" cannot contain leading or trailing white space, found %q`, chip)
+		}
+		if exists[chip] {
+			return fmt.Errorf(`"source-chip" cannot contain duplicate chip names, found %q`, chip)
+		}
+		exists[chip] = true
+	}
+	return nil
+}
+
+// XXX: What should be the limit on max range.
```

Comment:
@bboozzoo any suggestion here? is there a  limit in the kernel? otherwise we should just remove the XXX

Reply:
we can set it to 512, which would still be very high and unlikely to ever be reached in practice.